### ISSUE TITLE
Docker drops large HTTP response when LRO is turned on.

### ISSUE THREAD

**Issue by vhosakot** [docker/docker#32023]
**Description**

Large Receive Offload (LRO) is a feature that offloads processing of incoming packets to the NIC.
https://en.wikipedia.org/wiki/Large_receive_offload
LRO increases the throughput of high-bandwidth networks and reduces the CPU overhead as the NIC processes the incoming packets instead of the CPU. When LRO is turned on on the Docker host, the following commands do not work inside the container which pings www.google.com without any issue.
```
curl www.google.com
curl http://www.google.com
curl https://www.google.com
```
But `curl google.com` works fine inside the container. It looks like, when LRO is turned on on the Docker host, the Docker container drops curl's response when it is large and needs fragmentation.

When the following curl commands are run inside the container, the server (google) sends a large response that needs fragmentation:
```
curl www.google.com
curl http://www.google.com
curl https://www.google.com
```

`curl google.com` does not receive a large response that needs fragmentation. Hence, `curl google.com` works fine inside the container.

When LRO is turned off on the Docker host, all of the following curl commands work fine inside the container:
```
curl www.google.com
curl http://www.google.com
curl https://www.google.com
curl google.com
```

LRO is a per-interface setting. If LRO is turned on for an interface, the CPU will offload the processing of incoming packets (received on that interface) to the NIC. To reproduce this bug, LRO must be turned on for the public/external interface on the Docker host which reaches www.google.com.

- Docker host is a bare-metal server (HP DL380 superdome) running RHEL 7.3.
- Docker container is a CentOS container.
- All interfaces on both the docker host and inside the CentOS container are using the default MTU value of 1500 bytes.

**Steps to reproduce the issue:**
1. Install docker on docker host (`yum -y install docker`)
2. Ping www.google.com from the docker host (bare-metal server HP DL380 superdome in this case). Note the IP address of www.google.com in the output of ping (`172.217.6.68` in the example below).
```
# ping -c 2 www.google.com
PING www.google.com (172.217.6.68) 56(84) bytes of data.
64 bytes from sfo07s17-in-f68.1e100.net (172.217.6.68): icmp_seq=1 ttl=56 time=10.2 ms
64 bytes from sfo07s17-in-f68.1e100.net (172.217.6.68): icmp_seq=2 ttl=56 time=10.2 ms
--- www.google.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 10.221/10.225/10.229/0.004 ms
```
3. Find the interface on the docker host that reaches www.google.com (`eth0` in the example below).
```
# ip route get 172.217.6.68
172.217.6.68 via 192.168.9.1 dev eth0  src 192.168.9.160 
    cache 
```
4. Turn on LRO for the interface on the docker host that reaches www.google.com (`eth0` in the example below).
```
ethtool -K eth0 lro on
```
5. Make sure that LRO is turned on for the interface on the docker host that reaches www.google.com (`eth0` in the example below).
```
# ethtool -k eth0 | grep large-receive-offload
large-receive-offload: on
```
6. Run a CentOS container.
```
docker run -it --hostname test_container centos /bin/bash
```
7. Run the following commands inside the container. None of the commands below work. They just hang forever since they do not receive curl's response. The commands below make the server (google) send a large response that needs fragmentation:.
```
[root@test_container /]# curl www.google.com
[root@test_container /]# curl http://www.google.com
[root@test_container /]# curl https://www.google.com
```
8. Run the following command inside the container. It works fine as the command below does not receive large response that needs fragmentation.
```
[root@test_container /]# curl google.com
```
9. Turn off LRO for the interface on the docker host that reaches www.google.com (`eth0` in the example below).
```
ethtool -K eth0 lro off
```
10. Make sure that LRO is turned off for the interface on the docker host that reaches www.google.com (`eth0` in the example below).
```
# ethtool -k eth0 | grep large-receive-offload
large-receive-offload: off
```
11. Now, re-run the same commands in step 7 above. They work fine and do not hang. The received curl response will be seen in the output.
12. Turn on LRO again for `eth0` on the docker host (`ethtool -K eth0 lro on`)
13. Turn on packet capture (`tcpdump -i any port http or port https`) on the docker host
14. Re-run the commands in step 7 above. We can see the curl's large response come into the docker host in the output of tcpdump, but not into the CentOS container (looks like `docker0` dropped them).

**Describe the results you received:**

See step 7 above.

**Describe the results you expected:**

The commands in step 7 above must work like step 8.

**Output of `docker version`:**

```
# docker version
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:38:28 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:38:28 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
# docker info
Containers: 16
 Running: 15
 Paused: 0
 Stopped: 1
Images: 12
Server Version: 1.13.1
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 31.26 GiB
Name: mcp1.paslab009007.mc.metacloud.in
ID: 3TXM:6OBI:CYFH:CDDU:52B6:I6IH:J2F7:GDM6:W6KX:O5X5:ZH4H:AC6A
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Experimental: false
Insecure Registries:
 10.233.0.0/18
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

- Docker host is a bare-metal server (HP DL380 superdome) running RHEL 7.3.
- Docker container is a CentOS container.
- All interfaces on both the docker host and inside the CentOS container are using the default MTU value of 1500 bytes.

**Comment by vhosakot** [comment]
Adding @trinhlee and @romahony who saw this issue too.

**Comment by thaJeztah** [comment]
Thanks for a clear description of the issue 👍 

ping @aboch ptal

**Comment by cpuguy83** [comment]
Kinda looks similar to http://ehaselwanter.com/en/blog/2014/11/02/mtu-issue--nope-it-is-lro-with-bridge-and-bond/

**Comment by vhosakot** [comment]
@cpuguy83 thanks for the link!

http://ehaselwanter.com/en/blog/2014/11/02/mtu-issue--nope-it-is-lro-with-bridge-and-bond/ has the link http://downloadmirror.intel.com/22919/eng/README.txt which is Intel's official documentation that says:
```
WARNING:  The ixgbe driver compiles by default with the LRO (Large Receive
Offload) feature enabled.  This option offers the lowest CPU utilization for
receives, but is completely incompatible with *routing/ip forwarding* and
*bridging*.  If enabling ip forwarding or bridging is a requirement, it is
necessary to disable LRO using compile time options as noted in the LRO
section later in this document.  The result of not disabling LRO when combined
with ip forwarding or bridging can be low throughput or even a kernel panic.
```
Yes, this bug is seen with the `ixgbe` driver on the Docker host. So, does this mean LRO should not be turned on since it causes problems with the `docker0` bridge? If yes, when LRO is turned on, shouldn't docker flag an error when we use/start/run docker? Shouldn't this be noted in docker networking's documentation. Shouldn't docker check if LRO is turned off when starting?

I think docker should identify hardware features not allowed and flag them as errors when docker starts.

**Comment by justincormack** [comment]
It seems like a bug in the Intel driver that it ships with an option that disables normal usage of the network interface. If someone wants to add a check for this to Docker or to the `check-config` script that could be helpful.

**Comment by vhosakot** [comment]
@justincormack No, I don't think this is a bug in Intel's `ixgbe` driver at all. In fact, many customers turn on LRO in production to offload packet processing to hardware and to achieve higher network throughput.  

LRO has been used (and turned on) for ages now. I think Docker must handle LRO and any disallowed hardware offloading properly (with an error or a note in Docker's documentation).

**Comment by cpuguy83** [comment]
@vhosakot The bug is that forwarding, bridging, etc are not support with LRO and the driver is supposed to disable it for traffic for these cases.
Based on what I've read it already automatically disables LRO for macvlans and a couple of other things, but does not properly handle veth (bridge devices).

**Comment by SneakyMax** [comment]
For those that come across this, when using a Network Load Balancer on Amazon EC2, LRO is active, causing large requests to fail. It actually uses GRO, so the command to disable it is:

```bash
ethtool -K eth0 gro off
```

(This is still unfortunately a problem in mid-2019)

**Comment by ForterraIT** [comment]
Sorry to revive a necro post, but this issue still exists in 2024.
